File: data-structure/LinkedList/Python/singlyLL/SinglyLL.py
# A Node class which has a constructor
# accepts data and next pointer
class Node(object):
	def __init__(self, data, next):
		self.data = data
		self.next = next

class SinglyLL(object):
	# head and tail initially points to None
	head = None
	tail = None

	# Append data to LL
	def add(self, data):
		# Create a temporary node
		# If the head pointer is empty i.e LL is empty
		# Then head will be temprary node which has data
		# else add data to next to tail node

		node = Node(data, None) 
		if self.head is None:
		    self.head = self.tail = node
		else:
		    self.tail.next = node
		self.tail = node # IMP! and assign back that node to tail

	def add_front(self, data):
		node = Node(data, None)
		if self.head is None:
			self.head = node
		else:
			node.next = self.head
		self.head = node

	def remove(self):
		# go to the last node
		# and keep traking the prev node
		# then assign prev node's next to None

		if self.head is None:
			print "list is empty, can't remove"
			pass
		else:
			temp_node = self.head
			prev_node = None
			while temp_node.next is not None:
				prev_node = temp_node
				temp_node = temp_node.next
			prev_node.next = None

	def remove_front(self):
		# store head pointer to temp node
		# then temp node next to its next
		# and head to temp node (which has moved to next)

		temp_node = self.head
		if self.head is None:
			print "list is empty, can't remove"
			pass
		else:
			temp_node = temp_node.next
			self.head = temp_node

	def display(self):
		temp_node = self.head
		while temp_node is not None:
		    print temp_node.data, " -> ",
		    temp_node = temp_node.next
		print None

File: data-structure/LinkedList/SinglyLL/Javascript/SinglyLL.js
var Node = function(data) {
	this.data = data;
	this.next = null;
};

var SinglyLL = function() {
	this.len = 0;
	this.head = null;
	this.tail = null;
};

SinglyLL.prototype.add = function(data) {
	var newNode = new Node(data);
	// Empty LL
	if(this.head === null) {
		this.head = this.tail = newNode;
		this.len = 1;
	} else 
		this.tail.next = newNode;
	this.tail = newNode;
	this.len++;
};

SinglyLL.prototype.addFront = function(data) {
	// 1. create new node
	// 2. Assign new node next to the current head
	// 3. Then new node to head

	var newNode = new Node(data);
	newNode.next = this.head;
	this.head = newNode;
};

SinglyLL.prototype.remove = function() {
	// 1. Check for empty node
	// 2. Go up to the last node
	// 3. Keep track the prev node
	// 4. Assign next of prev to null
	// 5. Output the deleted node

	if(this.head == null) {
		console.log("cant remove! Linked List is empty.");
	} else {
		var temp = this.head;
		var prev = null;
		while(temp.next !== null) {
			prev = temp;
			temp = temp.next;
		}
		prev.next = null;
	}

};

SinglyLL.prototype.display = function() {
	var temp = this.head;
	var output ='';
	while(temp !== null) {
		output = output + temp.data + " -->";
		temp = temp.next;
	}
	console.log(output + " null");
};

var list = new SinglyLL();
list.add(1);
list.add(2);
list.add(3);
list.add(4);
list.addFront(100);
list.display();
list.remove()
list.display();